### PR TITLE
edgeql: Verify that operators and implicit casts work.

### PR DESCRIPTION
**Pull request by vpetrovykh**
Add exhaustive tests for all valid scalar types and all defined
operators except `??` and `IF...ELSE` (these 2 operators have a
different implementation from the others). The tests validate that
the operators are legal between specific scalar types and that the
return type is also inferred correctly.

### CODE REVIEW

File: edb/lib/std/25-numoperators.eql [hunk 1]
```diff
@@ -934,19 +1038,29 @@ std::`%` (n: std::decimal, d: std::decimal) -> std::decimal
...
 
+# need an explicit operator for int64 in order to guarantee the result
+# is float64 and not decimal
 CREATE INFIX OPERATOR
-std::`^` (n: std::float64, p: std::float64) -> std::float64
-    FROM SQL OPERATOR '^';
+std::`^` (n: std::int64, p: std::int64) -> std::float64
+{
+    FROM SQL OPERATOR r'*';
+    FROM SQL 'SELECT (n ^ p)';
```
Comment (elprans) [inline edb/lib/std/25-numoperators.eql]: Shouldn't you be casing `p` to `float8` here to achieve the desired effect?

File: edb/lib/std/25-numoperators.eql [hunk 2]
```diff
@@ -300,12 +318,22 @@ std::`?!=` (l: OPTIONAL std::float64, r: OPTIONAL std::float64) -> std::bool
...
 
 CREATE INFIX OPERATOR
-std::`!=` (l: std::decimal, r: std::decimal) -> std::bool
+std::`!=` (l: std::decimal, r: std::anyreal) -> std::bool
+    FROM SQL OPERATOR r'<>';
+
+
+CREATE INFIX OPERATOR
+std::`?!=` (l: OPTIONAL std::decimal, r: OPTIONAL std::anyreal) -> std::bool
+    FROM SQL EXPRESSION;
```
Comment (1st1) [inline edb/lib/std/25-numoperators.eql]: Please use this style consistently:

```
{
     FROM SQL $$
     SELECT round("val")
     $$;
 };
```

File: edb/lib/std/25-numoperators.eql [hunk 3]
```diff
@@ -378,14 +406,33 @@ std::`>` (l: std::float64, r: std::float64) -> std::bool
...
-std::`>` (l: std::anyreal, r: std::decimal) -> std::bool
+std::`>` (l: std::decimal, r: std::decimal) -> std::bool
+    FROM SQL OPERATOR r'>';
+
+
+CREATE INFIX OPERATOR
+std::`>` (l: std::float64, r: std::decimal) -> std::bool
+    FROM SQL $$
+        SELECT "l" > "r";
+    $$;
```
Comment (1st1) [inline edb/lib/std/25-numoperators.eql]: Please use this formatting style:

```
{
     FROM SQL $$
     SELECT round("val")
     $$;
 };
```

File: edb/lib/std/25-numoperators.eql [hunk 4]
```diff
@@ -934,19 +1038,29 @@ std::`%` (n: std::decimal, d: std::decimal) -> std::decimal
...
+    FROM SQL OPERATOR r'*';
+    FROM SQL 'SELECT (n ^ p)';
+};
 
 
 CREATE INFIX OPERATOR
-std::`^` (n: std::int64, p: std::int64) -> std::decimal
+std::`^` (n: std::float32, p: std::float32) -> std::float32
 {
     FROM SQL OPERATOR '*';
```
Comment (1st1) [inline edb/lib/std/25-numoperators.eql]: Need a comment here clarifying why `^` is created from `*`.

File: tests/test_edgeql_calls.py [hunk 5]
```diff
@@ -1141,3 +1141,137 @@ class TestEdgeQLFuncCalls(tb.QueryTestCase):
...
+        The operator IS is normally static, so special care needs to
+        be taken when it's used inside polymorphic functions.
+    ''')
+    async def test_edgeql_calls_33(self):
+        await self.con.execute('''
+            CREATE FUNCTION test::call33(
+                a: anytype
+            ) -> bool
+                FROM EdgeQL $$
+                    SELECT a IS float32
```
Comment (1st1) [inline tests/test_edgeql_calls.py]: We should just prohibit using `IS` in polymorphic functions on polymorphic types.  We have no ETA when (if ever) we can add support for this.  So instead of xfail, please make it raise a proper exception.

File: tests/test_edgeql_calls.py [hunk 6]
```diff
@@ -1141,3 +1141,137 @@ class TestEdgeQLFuncCalls(tb.QueryTestCase):
...
+
+                DROP FUNCTION test::call33(
+                    a: anyreal
+                );
+            ''')
+
+    @test.xfail('''
+        All of the results are False, whereas some are expected to be
+        True because the operator '^' should be polymorphic.
+    ''')
```
Comment (1st1) [inline tests/test_edgeql_calls.py]: ?

Testing the return type of `^` operator can be done without putting it into a polymorphic function.  The purpose of this test is to show that `float32 ^ float32` is `float32`, right?  AFAIK you can fix this by adding a specialized `^` for `float32` and cast the result there.

File: tests/test_edgeql_expressions.py [hunk 7]
```diff
@@ -380,6 +381,574 @@ class TestExpressions(tb.QueryTestCase):
...
+        '<float32>1',
+        '<float64>1',
+        # decimal
+        '<decimal>1',
+    ]
+    unorderable = scalar_samples[:4]
+    orderable = scalar_samples[4:-6]
+    datetime = scalar_samples[5:-6]
+    numeric = scalar_samples[-6:]
+    nonnumeric = scalar_samples[:-6]
```
Comment (1st1) [inline tests/test_edgeql_expressions.py]: ¯\_(ツ)_/¯ Don't do this.  Refactor as follows:

```
unorderable = [
        '<bool>True',
         '<uuid>"d4288330-eea3-11e8-bc5f-7faf132b1d84"',
         '<bytes>b"Hello"',
         '<json>"Hello"',
]

orderable = [
		         '<str>"Hello"',
         '<datetime>"2018-05-07T20:01:22.306916+00:00"',
         '<naive_datetime>"2018-05-07T20:01:22.306916"',
         '<naive_date>"2018-05-07"',
         '<naive_time>"20:01:22.306916"',
         '<timedelta>"20:01:22.306916"',
]

...

scalar_samples = unorderable + orderable + ...
```

File: tests/test_edgeql_functions.py [hunk 8]
```diff
@@ -2254,9 +2267,21 @@ class TestEdgeQLFunctions(tb.QueryTestCase):
...
+    @unittest.skip('''
+        The cast distance of int16 is 2 for mean(int64) and
+        mean(float64). Currently instead of an error one of the
+        implementaitons will be used at random, so the test failure is
+        unstable.
+
+        Once the cast distance ambiguity is fixed the next issue will
+        be making sure that only decimals produce decimal result,
+        everything else is float32 or float64.
+    ''')
```
Comment (1st1) [inline tests/test_edgeql_functions.py]: Just fix this -- `compiler/func.py` lines 96..110.

File: edb/lib/std/25-numoperators.eql (hunk 1 above)
Comment (vpetrovykh) [reply edb/lib/std/25-numoperators.eql]: It is my understanding that at the level of SQL operators we don't want to have unnecessary casts. As far as I can tell (albeit based on the signature of `pow` and my limited tests with `maxint64 ^ some_integer_power`) PostgreSQL actually returns `float8` for two integer inputs.

A possible test of the underlying implementation might be `10^309` which is greater than maxfloat64 and should result in an overflow error, while `10^308` is still a `float64`. If this test sounds reasonable, I can add it.

I can add the above comment about PostgreSQL to that specific definition, too.

File: edb/lib/std/25-numoperators.eql (hunk 4 above)
Comment (vpetrovykh) [reply edb/lib/std/25-numoperators.eql]: Right, and I should probably copy `/` instead of the `*` anyways because power behaves more like division than multiplication w.r.t. types and other meta-information. This is simply because `2^-2` is `1/4`.

File: tests/test_edgeql_expressions.py (hunk 7 above)
Comment (vpetrovykh) [reply tests/test_edgeql_expressions.py]: Yeah, this is a historical artifact of how that testing set came to be. I'll refactor for the sake of sanity and clarity.

File: edb/lib/std/25-numoperators.eql (hunk 3 above)
Comment (vpetrovykh) [reply edb/lib/std/25-numoperators.eql]: I've updated indentation to be consistent.

File: edb/lib/std/25-numoperators.eql (hunk 2 above)
Comment (vpetrovykh) [reply edb/lib/std/25-numoperators.eql]: We consistently omit `{...}` for one-liners like `FROM SQL EXPRESSION`.

File: tests/test_edgeql_functions.py (hunk 8 above)
Comment (vpetrovykh) [reply tests/test_edgeql_functions.py]: Currently, a "fix" for ambiguity will result in breaking many tests (such as `ln(<int16>10)` being unable to choose between `int64` and `float64` implementations). Fixing that, will backfire on how `float32` + `int16` works, namely it will result in a `float64` or it will require explicit versions for every numeric scalar of every std and math function. We also don't yet have validation that the SQL function body returns what is promised in EdgeQL. All of this sounds like a whole separate PR that's concentrating on *functions*.

The bottom line is that this fix is not in the scope of this PR, but defining sane *implicit casts* and *operators* is. So it seems that much like with `decimal` we need to adopt a specific doctrine w.r.t. `float32`. Just how special is it? How contagious? And why? Similarly, how special are `int16` and `int32`?

File: tests/test_edgeql_calls.py (hunk 6 above)
Comment (vpetrovykh) [reply tests/test_edgeql_calls.py]: Dropping the test because polymorphic functions will not be supported in user-space.

File: tests/test_edgeql_calls.py (hunk 5 above)
Comment (vpetrovykh) [reply tests/test_edgeql_calls.py]: I'll make a separate small PR forbidding `IS` with polymorphic types.

**Comment by vpetrovykh** [comment]
I've added tests for `??` and `IF/ELSE`.

File: edb/lang/edgeql/compiler/expr.py [hunk 9]
```diff
@@ -356,6 +356,16 @@ def compile_IfElse(
...
     else_expr_type = inference.infer_type(else_expr, ctx.env)
+    cond_expr_type = inference.infer_type(condition, ctx.env)
+
+    # make sure that the condition is actually boolean
+    bool_t = ctx.env.schema.get('std::bool')
+    if not cond_expr_type.issubclass(ctx.env.schema, bool_t):
+        raise errors.EdgeQLError(
+            'if/else condition must be of type {}, got: {}'.format(
+                bool_t.get_name(ctx.env.schema),
+                cond_expr_type.get_name(ctx.env.schema)),
```
Comment (1st1) [inline edb/lang/edgeql/compiler/expr.py]: `.get_name()` -> `.get_displayname()`

File: edb/lib/std/25-numoperators.eql [hunk 10]
```diff
@@ -24,6 +24,14 @@
 # for each permutation of common integer and floating-point
 # operand types to avoid casting overhead, as these operations
 # are very common.
+#
+# Our implicit casts do not coincide with PostgreSQL. In particular we
+# do not implicitly cast floats into decimals. The philosophy behind
+# that is that using decimal arithmetic should be opt-in. One of the
```
Comment (1st1) [inline edb/lib/std/25-numoperators.eql]: "decimal arithmetic should be opt-in." and that once you're doing decimal arithmetic you won't accidentally switch to doing floating point arithmetic.

File: edb/lib/std/25-numoperators.eql [hunk 11]
```diff
@@ -930,23 +1034,43 @@ CREATE INFIX OPERATOR
 std::`%` (n: std::decimal, d: std::decimal) -> std::decimal
 {
     FROM SQL OPERATOR r'%';
-    FROM SQL 'SELECT n - floor(n / d) * d';
+    FROM SQL 'SELECT n - floor("n" / "d") * d';
```
Comment (1st1) [inline edb/lib/std/25-numoperators.eql]: `d` -> `"d"`

File: edb/lib/std/25-numoperators.eql [hunk 12]
```diff
@@ -930,23 +1034,43 @@ CREATE INFIX OPERATOR
...
 
 
+# need an explicit operator for int64 in order to guarantee the result
+# is float64 and not decimal
 CREATE INFIX OPERATOR
-std::`^` (n: std::float64, p: std::float64) -> std::float64
-    FROM SQL OPERATOR '^';
+std::`^` (n: std::int64, p: std::int64) -> std::float64
+{
+    # We cheat here a bit by copying the `/` operator, since there is
```
Comment (1st1) [inline edb/lib/std/25-numoperators.eql]: "by copying the `/` operator" -> "by copying most of SQL operator metadata from the `/` SQL operator"

File: tests/test_edgeql_functions.py [hunk 13]
```diff
@@ -2038,9 +2039,21 @@ class TestEdgeQLFunctions(tb.QueryTestCase):
             {50000000000000000000000000},
         ])
 
+    @unittest.skip('''
```
Comment (1st1) [inline tests/test_edgeql_functions.py]: Please add "edb.server._testbase.skip()" decorator (just alias it to unittest.skip)

File: edb/lang/edgeql/compiler/expr.py (hunk 9 above)
Comment (vpetrovykh) [reply edb/lang/edgeql/compiler/expr.py]: Replaced in that file.

File: edb/lib/std/25-numoperators.eql (hunk 10 above)
Comment (vpetrovykh) [reply edb/lib/std/25-numoperators.eql]: Added the clarification that decimals should not be accidentally switched back to floats.

File: edb/lib/std/25-numoperators.eql (hunk 11 above)
Comment (vpetrovykh) [reply edb/lib/std/25-numoperators.eql]: Corrected.

File: edb/lib/std/25-numoperators.eql (hunk 12 above)
Comment (vpetrovykh) [reply edb/lib/std/25-numoperators.eql]: Replaced this comment with a slightly longer more explicit version in this file.

File: tests/test_edgeql_functions.py (hunk 13 above)
Comment (vpetrovykh) [reply tests/test_edgeql_functions.py]: Added an alias to the "decorators.py".
